### 🔄 Actualización incremental de artículos desde PubMed

#### 🌍 Contexto

Una vez que tenemos artículos científicos indexados en nuestro sistema, surge una pregunta clave: **¿cómo mantenemos la información actualizada?**

PubMed no es una base de datos estática. Cada día se añaden nuevos artículos, se revisan registros existentes (correcciones de autores, abstracts actualizados, nuevos metadatos) y, ocasionalmente, se eliminan registros. Si no actualizamos nuestro sistema, el buscador quedará obsoleto con el tiempo.

El reto está en hacerlo de forma eficiente: con más de 40 millones de artículos en PubMed, no podemos revisar todo el catálogo cada vez que queramos actualizar.

La solución es usar los **daily update files** — archivos que NLM (la institución que gestiona PubMed) publica diariamente con exactamente los cambios del día: qué es nuevo, qué ha cambiado y qué se ha eliminado.

---

#### 🎯 Objetivo

Validar y diseñar una estrategia de **actualización incremental** que, en lugar de reindexar todo el corpus, procese únicamente los artículos afectados por los cambios del día.

---

#### ⚙️ Estrategia

Comparamos dos enfoques:

- **Estrategia antigua (proceso escoba)**: revisar todos los artículos ya indexados para comprobar si han cambiado. Funciona, pero se vuelve cada vez más lento a medida que el corpus crece.
- **Estrategia nueva (update files)**: descargar el fichero de cambios diario de NLM y cruzarlo con nuestra colección. Solo procesamos los artículos que realmente han cambiado.

---

#### 🧩 Pasos que seguimos

##### 1. Exploración del formato de los update files
Entendemos cómo están estructurados los archivos de actualización de PubMed: qué contienen, cómo se llaman y cómo acceder a ellos.

##### 2. Extracción de los tres conjuntos de cambios
De cada fichero extraemos:
- 🟢 **PMIDs nuevos** — artículos que aparecen por primera vez
- 🟡 **PMIDs revisados** — artículos que ya existían pero han cambiado
- 🔴 **PMIDs eliminados** — artículos que deben borrarse del sistema

##### 3. Cruce con nuestra colección
Comparamos los PMIDs del fichero con los que ya tenemos indexados en nuestra base de datos vectorial (Qdrant):
- Revisado + existe en nuestra colección → actualizar (*upsert*)
- Eliminado + existe en nuestra colección → borrar
- Nuevo + no existe → aplicar filtro temático e insertar si es relevante

##### 4. Validación de la estrategia
Comprobamos que el enfoque funciona correctamente antes de integrarlo en el pipeline de producción.

---

> **🚀 Qué NO hacemos aquí**
> Este notebook valida la lógica de detección de cambios. La descarga del contenido actualizado, la regeneración de embeddings y el upsert en Qdrant se integran en el pipeline principal.
>
> **💡 Idea clave**
> - Con los update files minimizamos el procesamiento requerido para la ctualización, ya que, en lugar de consultar el total de los documentos ya indexados, el sistema solo reevalúa el subconjunto afectado por los update files publicados por PubMed. 
>
> - Los registros clasificados como nuevos en los update files no se incorporan automáticamente. Su inclusión depende de que satisfagan los criterios de recuperación definidos en el proceso de ingesta (nuestra base de datos vectorial actúa como fuente de verdad). En cambio, los registros revisados o eliminados sí pueden cruzarse directamente con los PMIDs ya indexados para ejecutar operaciones de actualización o borrado selectivo. 

In [ ]:
# 1) Descargar o cargar un daily update file de PubMed
# 2) Parsear el XML
# 3) Extraer:
#    - PMIDs nuevos
#    - PMIDs revisados
#    - PMIDs borrados
# 4) Cruzarlos con una muestra de PMIDs de Qdrant
# 5) Medir cuántos irían a:
#    - insert
#    - refresh
#    - delete

In [3]:
# Importamos librerias y cargamos API key
import os
from dotenv import load_dotenv
import requests
from xml.etree import ElementTree as ET

import time
import pandas as pd
import torch
import matplotlib.pyplot as plt

from sentence_transformers import SentenceTransformer, util
from IPython.display import display

from __future__ import annotations

import gzip
import io
import re
from pathlib import Path
from typing import Iterable, Dict, Set, Tuple, List, Optional
import requests
from xml.etree import ElementTree as ET

In [2]:
load_dotenv()
ncbi_api_key = os.getenv('NCBI_API_KEY')
print(f"API Key cargada." if ncbi_api_key else "API Key no encontrada")

API Key cargada.


---

Los update files tiene un formato como el siquiente: 

pubmed25n1301.xml.gz    
pubmed25n1302.xml.gz   
pubmed25n1303.xml.gz   
...   

El nombre se construye:  pubmedYYnXXXX.xml.gz   
  
Donde:    
- YY → año (25 = 2025)
- XXXX → número secuencial del fichero

---
 
En producción habría que listar el directorio ftp con: https://ftp.ncbi.nlm.nih.gov/pubmed/updatefiles/   
Guardar las_processed_file y a partir de ahí procesar el siguiente...   


#### 🧩 Función: listar update files (ordenados)

In [7]:
import requests
import re
from urllib.parse import urljoin


def list_pubmed_update_files(config: dict):
    """
    Lista los archivos updatefiles de PubMed y devuelve los más recientes.

    Params (config dict):
    - base_url: str
    - limit: int (cuántos últimos devolver)

    Returns:
    - list[dict]: [{"filename": ..., "url": ...}]
    """

    base_url = config.get(
        "base_url",
        "https://ftp.ncbi.nlm.nih.gov/pubmed/updatefiles/"
    )
    limit = config.get("limit", 10)

    response = requests.get(base_url, timeout=30)
    response.raise_for_status()

    html = response.text

    # Buscar archivos tipo pubmedYYnXXXX.xml.gz
    pattern = r'pubmed\d+n\d+\.xml\.gz'
    files = list(set(re.findall(pattern, html)))

    # Ordenar por número secuencial
    def extract_number(filename):
        match = re.search(r'n(\d+)', filename)
        return int(match.group(1)) if match else 0

    files_sorted = sorted(files, key=extract_number)

    latest_files = files_sorted[-limit:]

    return [
        {
            "filename": f,
            "url": urljoin(base_url, f)
        }
        for f in latest_files
    ]

In [8]:
config = {
    "limit": 5
}

latest_files = list_pubmed_update_files(config)

for f in latest_files:
    print(f)

{'filename': 'pubmed26n1388.xml.gz', 'url': 'https://ftp.ncbi.nlm.nih.gov/pubmed/updatefiles/pubmed26n1388.xml.gz'}
{'filename': 'pubmed26n1389.xml.gz', 'url': 'https://ftp.ncbi.nlm.nih.gov/pubmed/updatefiles/pubmed26n1389.xml.gz'}
{'filename': 'pubmed26n1390.xml.gz', 'url': 'https://ftp.ncbi.nlm.nih.gov/pubmed/updatefiles/pubmed26n1390.xml.gz'}
{'filename': 'pubmed26n1391.xml.gz', 'url': 'https://ftp.ncbi.nlm.nih.gov/pubmed/updatefiles/pubmed26n1391.xml.gz'}
{'filename': 'pubmed26n1392.xml.gz', 'url': 'https://ftp.ncbi.nlm.nih.gov/pubmed/updatefiles/pubmed26n1392.xml.gz'}


In [12]:
import gzip
import requests
import xml.etree.ElementTree as ET


def inspect_pubmed_update_file_structure(config: dict) -> dict:
    url = config["url"]
    timeout = config.get("timeout", 60)
    first_n_children = config.get("first_n_children", 10)

    print(f"Descargando: {url}")

    response = requests.get(url, timeout=timeout)
    response.raise_for_status()

    print("Status:", response.status_code)
    print("Bytes comprimidos:", len(response.content))

    xml_bytes = gzip.decompress(response.content)
    print("Bytes descomprimidos:", len(xml_bytes))

    root = ET.fromstring(xml_bytes)
    print("XML parseado correctamente")

    first_children_tags = [child.tag for child in list(root)[:first_n_children]]

    pubmed_article_count = len(root.findall(".//PubmedArticle"))
    delete_citation_count = len(root.findall(".//DeleteCitation"))

    summary = {
        "url": url,
        "root_tag": root.tag,
        "first_children_tags": first_children_tags,
        "pubmed_article_count": pubmed_article_count,
        "delete_citation_count": delete_citation_count,
        "total_top_level_children": len(list(root)),
    }

    return summary

In [13]:
file_to_inspect = latest_files[-1]

config = {
    "url": file_to_inspect["url"],
    "timeout": 60,
    "first_n_children": 10
}

summary = inspect_pubmed_update_file_structure(config)

print(summary)

Descargando: https://ftp.ncbi.nlm.nih.gov/pubmed/updatefiles/pubmed26n1392.xml.gz
Status: 200
Bytes comprimidos: 37534255
Bytes descomprimidos: 211041387
XML parseado correctamente
{'url': 'https://ftp.ncbi.nlm.nih.gov/pubmed/updatefiles/pubmed26n1392.xml.gz', 'root_tag': 'PubmedArticleSet', 'first_children_tags': ['PubmedArticle', 'PubmedArticle', 'PubmedArticle', 'PubmedArticle', 'PubmedArticle', 'PubmedArticle', 'PubmedArticle', 'PubmedArticle', 'PubmedArticle', 'PubmedArticle'], 'pubmed_article_count': 13607, 'delete_citation_count': 1, 'total_top_level_children': 13608}


### Paso 1: inspeccionar los primeros hijos del root

In [14]:
import gzip
import requests
import xml.etree.ElementTree as ET


def inspect_top_level_nodes(config: dict) -> None:
    url = config["url"]
    timeout = config.get("timeout", 60)
    n = config.get("n", 5)

    response = requests.get(url, timeout=timeout)
    response.raise_for_status()

    xml_bytes = gzip.decompress(response.content)
    root = ET.fromstring(xml_bytes)

    print(f"Root tag: {root.tag}")
    print(f"Total top-level children: {len(list(root))}")
    print("-" * 80)

    for i, child in enumerate(list(root)[:n]):
        print(f"[{i}] tag = {child.tag}")
        print(f"    atributos = {child.attrib}")
        print(f"    nº hijos inmediatos = {len(list(child))}")
        print(f"    hijos inmediatos = {[c.tag for c in list(child)[:10]]}")
        print("-" * 80)

In [23]:
config = {
    "url": latest_files[-1]["url"],
    "timeout": 60,
    "n": 4
}

inspect_top_level_nodes(config)

Root tag: PubmedArticleSet
Total top-level children: 13608
--------------------------------------------------------------------------------
[0] tag = PubmedArticle
    atributos = {}
    nº hijos inmediatos = 2
    hijos inmediatos = ['MedlineCitation', 'PubmedData']
--------------------------------------------------------------------------------
[1] tag = PubmedArticle
    atributos = {}
    nº hijos inmediatos = 2
    hijos inmediatos = ['MedlineCitation', 'PubmedData']
--------------------------------------------------------------------------------
[2] tag = PubmedArticle
    atributos = {}
    nº hijos inmediatos = 2
    hijos inmediatos = ['MedlineCitation', 'PubmedData']
--------------------------------------------------------------------------------
[3] tag = PubmedArticle
    atributos = {}
    nº hijos inmediatos = 2
    hijos inmediatos = ['MedlineCitation', 'PubmedData']
--------------------------------------------------------------------------------


 #### Observaciones
 El update file tiene como raíz `<PubmedArticleSet>`
   
 A nivel top-level aparecen dos tipos de nodos:
 - PubmedArticle
 - DeleteCitation

### Paso 2: inspeccionar un nodo completo 

In [16]:
def preview_first_node_of_type(config: dict) -> None:
    url = config["url"]
    node_tag = config["node_tag"]   # "PubmedArticle" o "DeleteCitation"
    timeout = config.get("timeout", 60)
    preview_chars = config.get("preview_chars", 3000)

    response = requests.get(url, timeout=timeout)
    response.raise_for_status()

    xml_bytes = gzip.decompress(response.content)
    root = ET.fromstring(xml_bytes)

    node = root.find(f"./{node_tag}")

    if node is None:
        print(f"No se encontró ningún nodo {node_tag}")
        return

    xml_str = ET.tostring(node, encoding="unicode")
    print(xml_str[:preview_chars])

#### Nodo PubmedArticle

In [46]:
config = {
    "url": latest_files[-1]["url"],
    "node_tag": "PubmedArticle",
    "preview_chars": 8000
}

preview_first_node_of_type(config)

<PubmedArticle>
    <MedlineCitation Status="MEDLINE" IndexingMethod="Manual" Owner="NLM">
      <PMID Version="1">21451503</PMID>
      <DateCompleted>
        <Year>2011</Year>
        <Month>08</Month>
        <Day>17</Day>
      </DateCompleted>
      <DateRevised>
        <Year>2026</Year>
        <Month>03</Month>
        <Day>23</Day>
      </DateRevised>
      <Article PubModel="Print-Electronic">
        <Journal>
          <ISSN IssnType="Electronic">1935-3456</ISSN>
          <JournalIssue CitedMedium="Internet">
            <Volume>4</Volume>
            <Issue>3</Issue>
            <PubDate>
              <Year>2011</Year>
              <Month>May</Month>
            </PubDate>
          </JournalIssue>
          <Title>Mucosal immunology</Title>
          <ISOAbbreviation>Mucosal Immunol</ISOAbbreviation>
        </Journal>
        <ArticleTitle>Initiation and regulation of T-cell responses in tuberculosis.</ArticleTitle>
        <Pagination>
          <MedlinePgn>288-93<

#### Nodo DeleteCitation

In [18]:
config = {
    "url": latest_files[-1]["url"],
    "node_tag": "DeleteCitation",
    "preview_chars": 2000
}

preview_first_node_of_type(config)

<DeleteCitation>
<PMID Version="1">41653471</PMID>
<PMID Version="1">41784435</PMID>
<PMID Version="1">41854130</PMID>
<PMID Version="1">41854507</PMID>
<PMID Version="1">41854589</PMID>
<PMID Version="1">41854593</PMID>
<PMID Version="1">41854622</PMID>
<PMID Version="1">41855335</PMID>
<PMID Version="1">41855343</PMID>
<PMID Version="1">41855549</PMID>
<PMID Version="1">41855551</PMID>
<PMID Version="1">41856362</PMID>
<PMID Version="1">41856963</PMID>
<PMID Version="1">41860156</PMID>
<PMID Version="1">41860984</PMID>
<PMID Version="1">41861152</PMID>
<PMID Version="1">41865249</PMID>
<PMID Version="1">41866141</PMID>
<PMID Version="1">41866144</PMID>
<PMID Version="1">41866154</PMID>
<PMID Version="1">41866165</PMID>
<PMID Version="1">41866177</PMID>
<PMID Version="1">41866324</PMID>
<PMID Version="1">41866528</PMID>
<PMID Version="1">41866529</PMID>
<PMID Version="1">41866660</PMID>
<PMID Version="1">41867160</PMID>
<PMID Version="1">41869790</PMID>
<PMID Version="1">41870286</PMI

#### Observaciones

PubmedArticle trae prácticamente el registro completo. Se parece mucho a EFetch XML, hasta el punto de que podríamos tratarlo como un registro reutilizable.

Un nodo DeleteCitation puede contener muchos PMIDs.

En la extracción de ejemplo, hay 1 nodo DeleteCitation y dentro de él, muchos PMID

#### Conclusiones 

A. PubmedArticle

Representa artículos presentes en ese update file.
Pueden corresponder a:
- artículos nuevos
- artículos revisados

B. DeleteCitation

Representa PMIDs que deben considerarse eliminados.
No trae el artículo completo, solo los identificadores a borrar.

Nos faltaría saber cómo separar automáticamente dentro de PubmedArticle cuáles son “new” y cuáles “revised”

### Paso 3: inspección estructural sin imprimir todo el XML

In [22]:
def print_node_tree(node: ET.Element, level: int = 0, max_depth: int = 3) -> None:
    indent = "  " * level
    print(f"{indent}- tag={node.tag}, attrib={node.attrib}")

    if level >= max_depth:
        return

    for child in list(node):
        print_node_tree(child, level + 1, max_depth=max_depth)


def inspect_first_node_tree(config: dict) -> None:
    url = config["url"]
    node_tag = config["node_tag"]
    timeout = config.get("timeout", 60)
    max_depth = config.get("max_depth", 3)

    response = requests.get(url, timeout=timeout)
    response.raise_for_status()

    xml_bytes = gzip.decompress(response.content)
    root = ET.fromstring(xml_bytes)

    node = root.find(f"./{node_tag}")

    if node is None:
        print(f"No se encontró ningún nodo {node_tag}")
        return

    print_node_tree(node, level=0, max_depth=max_depth)

In [20]:
config = {
    "url": latest_files[-1]["url"],
    "node_tag": "PubmedArticle",
    "max_depth": 3
}

inspect_first_node_tree(config)

- tag=PubmedArticle, attrib={}
  - tag=MedlineCitation, attrib={'Status': 'MEDLINE', 'IndexingMethod': 'Manual', 'Owner': 'NLM'}
    - tag=PMID, attrib={'Version': '1'}
    - tag=DateCompleted, attrib={}
      - tag=Year, attrib={}
      - tag=Month, attrib={}
      - tag=Day, attrib={}
    - tag=DateRevised, attrib={}
      - tag=Year, attrib={}
      - tag=Month, attrib={}
      - tag=Day, attrib={}
    - tag=Article, attrib={'PubModel': 'Print-Electronic'}
      - tag=Journal, attrib={}
      - tag=ArticleTitle, attrib={}
      - tag=Pagination, attrib={}
      - tag=ELocationID, attrib={'EIdType': 'doi', 'ValidYN': 'Y'}
      - tag=Abstract, attrib={}
      - tag=AuthorList, attrib={'CompleteYN': 'Y'}
      - tag=Language, attrib={}
      - tag=GrantList, attrib={'CompleteYN': 'Y'}
      - tag=PublicationTypeList, attrib={}
      - tag=ArticleDate, attrib={'DateType': 'Electronic'}
    - tag=MedlineJournalInfo, attrib={}
      - tag=Country, attrib={}
      - tag=MedlineTA, attrib=

In [21]:
config = {
    "url": latest_files[-1]["url"],
    "node_tag": "DeleteCitation",
    "max_depth": 3
}

inspect_first_node_tree(config)

- tag=DeleteCitation, attrib={}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attrib={'Version': '1'}
  - tag=PMID, attr

### Paso 4: Identificar archivos nuevos vs. revisados 

Esta función que recorre los primeros 50 PubmedArticle y busca/obtiene una tabla/resumen con:
- pmid
- pmid_version
- date_revised
- status
- atributos de MedlineCitation 

Así vemos si hay una señal clara para separar nuevos de revisados.  

In [26]:
import gzip
import requests
import xml.etree.ElementTree as ET


def inspect_pubmed_article_metadata(config: dict):
    url = config["url"]
    timeout = config.get("timeout", 60)
    limit = config.get("limit", 20)

    response = requests.get(url, timeout=timeout)
    response.raise_for_status()

    xml_bytes = gzip.decompress(response.content)
    root = ET.fromstring(xml_bytes)

    articles = root.findall("./PubmedArticle")[:limit]

    results = []

    for article in articles:
        medline = article.find("./MedlineCitation")
        if medline is None:
            continue

        pmid_node = medline.find("./PMID")
        date_revised = medline.find("./DateRevised")

        pmid = pmid_node.text.strip() if pmid_node is not None and pmid_node.text else None
        pmid_version = pmid_node.attrib.get("Version") if pmid_node is not None else None

        revised_str = None
        if date_revised is not None:
            year = date_revised.findtext("Year")
            month = date_revised.findtext("Month")
            day = date_revised.findtext("Day")
            revised_str = f"{year}-{month}-{day}"

        results.append({
            "pmid": pmid,
            "pmid_version": pmid_version,
            "medline_attrib": medline.attrib,
            "date_revised": revised_str,
        })

    return results

In [ ]:
config = {
    "url": latest_files[-1]["url"],
    "limit": 5
}

results = inspect_pubmed_article_metadata(config)

for row in results:
    print(row)

{'pmid': '21451503', 'pmid_version': '1', 'medline_attrib': {'Status': 'MEDLINE', 'IndexingMethod': 'Manual', 'Owner': 'NLM'}, 'date_revised': '2026-03-23'}
{'pmid': '21349660', 'pmid_version': '1', 'medline_attrib': {'Status': 'MEDLINE', 'IndexingMethod': 'Manual', 'Owner': 'NLM'}, 'date_revised': '2026-03-23'}
{'pmid': '16908116', 'pmid_version': '1', 'medline_attrib': {'Status': 'MEDLINE', 'IndexingMethod': 'Manual', 'Owner': 'NLM'}, 'date_revised': '2026-03-23'}
{'pmid': '21501798', 'pmid_version': '1', 'medline_attrib': {'Status': 'MEDLINE', 'IndexingMethod': 'Manual', 'Owner': 'NLM'}, 'date_revised': '2026-03-23'}
{'pmid': '21601849', 'pmid_version': '1', 'medline_attrib': {'Status': 'MEDLINE', 'IndexingMethod': 'Manual', 'Owner': 'NLM'}, 'date_revised': '2026-03-23'}
{'pmid': '17056638', 'pmid_version': '1', 'medline_attrib': {'Status': 'MEDLINE', 'IndexingMethod': 'Manual', 'Owner': 'NLM'}, 'date_revised': '2026-03-23'}
{'pmid': '18205173', 'pmid_version': '1', 'medline_attrib'

### Parseamos un XML completo para tener una visión global de la estructura del archivo 

Como con los resultados obtenidos no nos aclaramos, formateamos todos los campos disponibles de 1 xml y lo pasamos a dataframe para analizar.

In [28]:
import gzip
import requests
import xml.etree.ElementTree as ET
import pandas as pd


def parse_pubmed_updatefile_to_df(config: dict) -> pd.DataFrame:

    import gzip
    import requests
    import xml.etree.ElementTree as ET
    import pandas as pd

    url = config["url"]
    timeout = config.get("timeout", 60)

    response = requests.get(url, timeout=timeout)
    response.raise_for_status()

    xml_bytes = gzip.decompress(response.content)
    root = ET.fromstring(xml_bytes)

    rows = []

    for article in root.findall("./PubmedArticle"):
        medline = article.find("./MedlineCitation")
        pubmed_data = article.find("./PubmedData")

        if medline is None:
            continue

        pmid_node = medline.find("./PMID")
        date_revised_node = medline.find("./DateRevised")

        pmid = pmid_node.text.strip() if pmid_node is not None and pmid_node.text else None
        pmid_version = pmid_node.attrib.get("Version") if pmid_node is not None else None

        medline_version_id = medline.attrib.get("VersionID")
        status = medline.attrib.get("Status")
        indexing_method = medline.attrib.get("IndexingMethod")

        date_revised = None
        if date_revised_node is not None:
            y = date_revised_node.findtext("Year")
            m = date_revised_node.findtext("Month")
            d = date_revised_node.findtext("Day")
            if y and m and d:
                date_revised = f"{y}-{m}-{d}"

        has_abstract = False
        abstract_len = 0

        article_node = medline.find("./Article")
        if article_node is not None:
            abstract_node = article_node.find("./Abstract")
            if abstract_node is not None:
                texts = [ab.text for ab in abstract_node.findall("./AbstractText") if ab.text]
                if texts:
                    has_abstract = True
                    abstract_len = len(" ".join(texts))

        publication_status = None
        if pubmed_data is not None:
            publication_status = pubmed_data.findtext("./PublicationStatus")

        rows.append({
            "pmid": pmid,
            "pmid_version": pmid_version,
            "medline_version_id": medline_version_id,
            "status": status,
            "indexing_method": indexing_method,
            "date_revised": date_revised,
            "publication_status": publication_status,
            "has_abstract": has_abstract,
            "abstract_len": abstract_len,
        })

    return pd.DataFrame(rows)

In [ ]:
config = {
    "url": latest_files[-1]["url"],
    "timeout": 60
}

df_update = parse_pubmed_updatefile_to_df(config)

In [31]:
df_update.sample(4, random_state=42)

,pmid,pmid_version,medline_version_id,status,indexing_method,date_revised,publication_status,has_abstract,abstract_len
8620,41865517,1,None,Publisher,NaN,2026-03-22,aheadofprint,True,1173
5381,41640360,1,None,MEDLINE,Automated,2026-03-23,ppublish,True,1549
385,26129882,1,None,MEDLINE,Manual,2026-03-23,ppublish,True,1771
7144,41818943,1,None,MEDLINE,Automated,2026-03-23,ppublish,True,1723


In [ ]:
df_update["pmid_version"].value_counts()
# pmid_version siempre es 1

pmid_version
1    13607
Name: count, dtype: int64

In [47]:
df_update["date_revised"].value_counts()

date_revised
2026-03-23    11063
2026-03-22     2196
2026-03-17      348
Name: count, dtype: int64

In [ ]:
# Esto sí que es importante, vemos que realmente en PubMed podemos extraer artículos que se encuentran en fases tempranas de publicación 
# Decidimos centrarnos en aquellas publicaciones cuyo estatus (más o menos secuencial) es MEDLINE
df_medline = df_update[df_update["status"] == "MEDLINE"]
print(len(df_medline))  

6637


## Observaciones

Estas son las estadísticas del update file analizado: https://ftp.ncbi.nlm.nih.gov/pubmed/updatefiles/pubmed26n1392_stats.html 

Las estadisticas mostradas son: 

- MEDLINE New      1506   
- MEDLINE Revised  5131   
- MEDLINE Total    6637 → Es solo un subconjunto de esos 13607, concretamente coincide para los casos en que status == "MEDLINE"   
  
- Citation Set Total = 13607 → Es TODO lo que hay en el XML, es decir, todos los nodos `<PubmedArticle>`   
- Deletes = 29    


>
> El campo status indica en qué fase está el artículo dentro del proceso editorial de PubMed. La última fase corresponde con MEDLINE. 
> En este punto, retrocedemos para filtrar en la ingesta SOLO las publicaciones cuyo status = MEDLINE y de la misma forma, en los xml de update files, filtraremos por dicho campo status. 
>

---

#### Estados principales en PubMed (MedlineCitation Status)

El campo **Status** dentro de `MedlineCitation` indica el estado del artículo dentro del flujo editorial de PubMed. No todos los registros tienen el mismo nivel de calidad ni de completitud.


| Status | Significado |
| :--- | :--- |
| **Publisher** | 🟡 Datos enviados directamente por la editorial. Son preliminares y pueden cambiar. |
| **In-Process** | 🟡 Artículo en proceso de indexación por parte de PubMed. Aún no tiene todos los metadatos. |
| **In-Data-Review** | 🟡 Registro en fase de revisión interna. Puede contener inconsistencias o estar incompleto. |
| **PubMed-not-MEDLINE** | 🟡 Artículo presente en PubMed pero que no ha sido indexado como MEDLINE. |
| **MEDLINE** | ✅ Artículo completamente procesado, indexado y curado. Incluye metadatos completos y términos MeSH. |

---

#### Filtro por Status == MEDLINE

Motivos

1. **Calidad de los datos**
   * Tienen metadatos completos.
   * Han pasado procesos de validación.
   * Incluyen información estructurada fiable.

2. **Estabilidad**
   Los estados intermedios pueden cambiar con el tiempo, modificando título o abstract. Esto es crítico en sistemas con **embeddings**:
   * ⚠️ **Cambios en texto** = Embeddings inconsistentes.
   * ⚠️ **Recalculo:** Obliga a procesar el vector de nuevo.

3. **Reducción de ruido**
   * Evita artículos incompletos.
   * Reduce duplicidades potenciales.
   * Mejora la precisión en el *retrieval*.

4. **Coherencia en sistemas RAG**
   En arquitecturas de recuperación semántica, se necesita información consistente y estable. **MEDLINE** garantiza esa base sólida para los modelos.


## ✅ Conclusiones

Las publicaciones ingestadas en nuestra base de datos serán la **fuente de verdad** del sistema. A partir de ellas, definimos la siguiente política de actualización:

**Política de ingesta**
La ingesta se realizará de forma periódica (mensualmente), restringida siempre al contexto temático del proyecto y a artículos con estado de publicación estable (`status = MEDLINE`). Los artículos en estados intermedios (`publisher`, `in-process`) quedan excluidos por su inestabilidad — su contenido puede cambiar antes de ser indexado definitivamente.

**Seguimiento de update files**
Registraremos el último update file procesado. Los archivos tienen una **numeración secuencial organizada por año** (`pubmedYYnXXXX.xml.gz`), lo que permite establecer fácilmente qué ficheros quedan por procesar desde la última ejecución.

**Proceso de actualización**
Con cada update file se extraen los `pm_id`s de las publicaciones con `status = MEDLINE` y se cruzan contra los `pm_id`s de la colección en Qdrant para determinar qué operación aplicar:

- `pm_id` presente en Qdrant → **upsert** (actualización)
- `pm_id` en `DeleteCitation` y presente en Qdrant → **borrado selectivo**
- `pm_id` no presente en Qdrant → aplicar filtro temático → **inserción** si es relevante

Este proceso es paralelo e independiente de la ingesta periódica de nuevos artículos, y opera exclusivamente sobre el subconjunto afectado por cada fichero, lo que lo hace eficiente independientemente del volumen acumulado en la colección.